### ISSUE TITLE
HA installs - Exec shell and viewing logs is not working for local cluster.

### ISSUE THREAD

**Issue by sangeethah** [rancher/rancher#17028]
Rancher server version - v2.2.0-alpha3

Steps to reproduce the problem:
Create a HA setup.

Try to Exec shell/view logs for any of the pods in the local cluster.
shell get closed/disconnected as soon as it gets launched.

Same behavior is also noted for "Launch kubectl" shell.

**Comment by cjellick** [comment]
@sangeethah @jiaqiluo is this a regression new to v2.2?

**Comment by cjellick** [comment]
marking as p0 under the assumption that this is a new regression

**Comment by jiaqiluo** [comment]
@cjellick 
Update on 12/31:
The issue is re-producable on Rancher:v2.2.0-alpha3 as it describes in the ticket.
The issue is re-producable on **Rancher:master (12/31)**
The issue is  **NOT** re-producable on Rancher:v2.1.5-rc2

**Comment by sangeethah** [comment]
Also when I try to access local cluster using remote kubectl , I get the following error:
```
MacBook-Pro-3:.ssh sangeethahariharan$ kubectl exec -it halting-oyster-rancher-7c967456cd-5rx7q -c rancher "loglevel --set debug" -n cattle-system
error: unable to upgrade connection: error dialing backend: x509: certificate signed by unknown authority
MacBook-Pro-3:.ssh sangeethahariharan$ 
```

**Comment by daxmc99** [comment]
Verified that the Kubeconfig returned by rancher will not allow you to exec inside the pods.

**Comment by cjellick** [comment]
@sangeethah did you have valid certificates configured? If you are not using truly valid certs, you cant expect that to work. That is a separate issue from the shell in the UI not working.

@daxmc99 only focus on the shell in the UI not working, not downloading and using the kubeconfig

**Comment by daxmc99** [comment]
@cjellick I am seeing this issue on *master* (7c0dc586) for local installs (non-HA) as well.

**Comment by sangeethah** [comment]
@cjellick We are using the rancher generated certificates in this case.

Rancher server was deployed using this option:
```
helm install rancher-latest/rancher \
  --name rancher \
  --namespace cattle-system \
  --set hostname=<test.com>
```

**Comment by cjellick** [comment]
And you were able to reproduce that way @sangeethah ?

**Comment by jiaqiluo** [comment]
Failed to test on the latest master which should have the fix.

1. I cannot launch the `local` cluster's  kubectl from the UI, and it gave me `Closed Code: 1006`
<img width="1831" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/53599176-cede7f80-3b63-11e9-8477-8f38afada2dd.png">

2. I cannot execute shell or view logs of a pod, and it shows the same `Closed Code: 1006` error

3. The GET request gets a  `500 Internal Server Error` statue

4. I can use kubectl on my own machine the access to the cluster and exec into a pod

**Comment by jiaqiluo** [comment]
The bugfix is validated on master 3/6
Rancher Setup: HA install 
Steps are the same as the issue describes.

**Comment by DereMann** [comment]
Is this bug fixed? I am facing the same issue on version v2.2.8

**Comment by james-mcgoodwin** [comment]
We are also experiencing this shell disconnect issue. 

Environment (Server Side) -- 
Rancher2 Version: 2.2.8
Kubernetes Version: 1.14.6
Docker Version: 19.3.2
Host OS: Centos 7.7

Environment (Proxy Front End) -- 
Host OS: Centos 7.7
HAProxy Version: 1.9.8-1

Environment (Client Side) --
kubectl version: 1.16.0
Client OS: MacOS 10.14.6


Architecture: 
3 X Rancher Local Cluster Hosts (etcd/control/worker roles applied to each)
4 X other clusters, each cluster with: 
-- 3 X etcd/control role nodes
-- 3 X worker role nodes

2 X Haproxy VM's sit in front of the 3 rancher local cluster hosts, providing 443 ingress to the Rancher hostname (IE: https://rancher2.example.cloud)

Connections to the URL/443 and from kubectl/443 go through HAProxy VIP and is round-robin-load-balanced to one of the three backend rancher local cluster hosts. 

In short, we've built the HAProxy load balancers as described in this document: 
https://rancher.com/docs/rancher/v2.x/en/installation/ha/


This is what I experience when trying to use kubectl exec to get a shell in the local cluster rancher container. 

This same experience occurs on the other four clusters as well, but I use local rancher container to demonstrate below: 

[kube_shell_disconnect_output.txt](https://github.com/rancher/rancher/files/3677515/kube_shell_disconnect_output.txt)

**Comment by sdevang** [comment]
I had the same issue and I managed to fix it by increasing load balancer's timeout to 3600. I hope this might help someone else in future too.

**Comment by jzandbergen** [comment]
After upgrading rancher from 2.4.8 to 2.5.3 I am facing the same issue.
kubectl from my machine is working. Kubectl and logs from the UI are not working.

**Comment by graememross** [comment]
I am encountering the exact same as @jzandbergen, Interestingly I have two clusters connected to my rancher, a largish development cluster used by our developers where this is happening and a small tool cluster which is working perfectly.
Both clusters were created the exact same way on the same types of machines.

**Comment by jzandbergen** [comment]
@graememross Eventually we found the route to the kubernetes services network was not set. This caused a lot of issues.
We fixed it by manually adding the route:
```
ip route add 10.43.0.0/16 dev flannel.1
```

**Comment by graememross** [comment]
@jzandbergen thanks for responding :)

Can you clarify, where exactly you had to set the route
  In the rancher container
  On the host running rancher
  In the affected cluster

**Comment by jzandbergen** [comment]
> @jzandbergen thanks for responding :)
> 
> Can you clarify, where exactly you had to set the route
> In the rancher container
> On the host running rancher
> In the affected cluster

I'm sorry for the confusion this was on the kubernetes nodes.

**Comment by graememross** [comment]
@jzandbergen Ok thanks for that ... makes sense
I will check our nodes
